File: backend/src/main/java/com/exam/seating/service/StudentService.java
package com.exam.seating.service;

import com.exam.seating.dto.request.StudentRequestDto;
import com.exam.seating.dto.response.StudentResponseDto;

import java.util.List;

public interface StudentService {

    List<StudentResponseDto> getAllStudents();

    StudentResponseDto getStudentById(Long id);

    StudentResponseDto getStudentByRegisterNumber(String registerNumber);

    StudentResponseDto getMyProfile(String username);

    List<StudentResponseDto> searchStudents(String keyword);

    StudentResponseDto createStudent(StudentRequestDto dto);

    StudentResponseDto updateStudent(Long id, StudentRequestDto dto);

    void deleteStudent(Long id);

    com.exam.seating.dto.response.StudentImportSummaryDto importStudentsFromCsv(org.springframework.web.multipart.MultipartFile file);

    com.exam.seating.dto.response.StudentImportSummaryDto importStudentsFromCsv(org.springframework.web.multipart.MultipartFile file, Long examId);

    com.exam.seating.dto.response.ConcurrentBranchImportSummaryDto importConcurrentBranchStudents(
            List<org.springframework.web.multipart.MultipartFile> files,
            List<String> branches,
            java.time.LocalDate examDate,
            java.time.LocalTime startTime,
            java.time.LocalTime endTime,
            Boolean autoAllocateSeating
    );

    String getConcurrentBranchTemplate(String branch);
}

File: frontend/src/app/core/services/student.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../models/api-response.model';
import { Student, StudentRequest, StudentImportSummary, ConcurrentBranchImportSummary } from '../models/student.model';

@Injectable({
  providedIn: 'root'
})
export class StudentService {
  private apiUrl = `${environment.apiUrl}/students`;

  constructor(private http: HttpClient) {}

  getAllStudents(): Observable<ApiResponse<Student[]>> {
    return this.http.get<ApiResponse<Student[]>>(this.apiUrl);
  }

  getMyProfile(): Observable<ApiResponse<Student>> {
    return this.http.get<ApiResponse<Student>>(`${this.apiUrl}/me`);
  }

  getStudentById(id: number): Observable<ApiResponse<Student>> {
    return this.http.get<ApiResponse<Student>>(`${this.apiUrl}/${id}`);
  }

  getStudentByRegisterNumber(regNo: string): Observable<ApiResponse<Student>> {
    return this.http.get<ApiResponse<Student>>(`${this.apiUrl}/register/${encodeURIComponent(regNo)}`);
  }

  searchStudents(keyword: string): Observable<ApiResponse<Student[]>> {
    return this.http.get<ApiResponse<Student[]>>(`${this.apiUrl}/search`, {
      params: { keyword }
    });
  }

  createStudent(student: StudentRequest): Observable<ApiResponse<Student>> {
    return this.http.post<ApiResponse<Student>>(this.apiUrl, student);
  }

  updateStudent(id: number, student: StudentRequest): Observable<ApiResponse<Student>> {
    return this.http.put<ApiResponse<Student>>(`${this.apiUrl}/${id}`, student);
  }

  deleteStudent(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`);
  }

  importStudents(file: File, examId?: number): Observable<ApiResponse<StudentImportSummary>> {
    const formData = new FormData();
    formData.append('file', file);
    const params: any = {};
    if (examId) {
      params.examId = examId;
    }
    return this.http.post<ApiResponse<StudentImportSummary>>(
      `${this.apiUrl}/import`,
      formData,
      { params }
    );
  }

  downloadCsvTemplate(): Observable<Blob> {
    return this.http.get(`${this.apiUrl}/csv-template`, {
      responseType: 'blob'
    });
  }

  importConcurrentBranchStudents(
    files: File[],
    examDate: string,
    startTime: string,
    endTime: string,
    autoAllocateSeating: boolean = true,
    branches?: string[]
  ): Observable<ApiResponse<ConcurrentBranchImportSummary>> {
    const formData = new FormData();
    for (const file of files) {
      formData.append('files', file);
    }
    if (branches && branches.length > 0) {
      for (const branch of branches) {
        formData.append('branches', branch);
      }
    }
    const params: any = {
      examDate,
      startTime,
      endTime,
      autoAllocateSeating
    };
    return this.http.post<ApiResponse<ConcurrentBranchImportSummary>>(
      `${this.apiUrl}/import-concurrent-branches`,
      formData,
      { params }
    );
  }

  downloadConcurrentTemplate(branch: string = 'CSE'): Observable<Blob> {
    return this.http.get(`${this.apiUrl}/concurrent-template`, {
      params: { branch },
      responseType: 'blob'
    });
  }
}
